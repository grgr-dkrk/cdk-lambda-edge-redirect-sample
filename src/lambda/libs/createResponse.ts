import { CloudFrontResultResponse } from 'aws-lambda'

/**
 * create response to redirect
 * @param { string } target - URL for redirect to
 * @param { string } status - 3xx Status Code
 * @param { string } description - Status Description
 * @return { CloudFrontResultResponse } response
 */
export const createRedirectResponse = (
  target: string,
  // TODO: fix types
  status?: '301' | '302' | '307',
  description?: string,
): CloudFrontResultResponse => ({
  status: status ?? '301',
  statusDescription: description,
  headers: {
    location: [
      {
        key: 'Location',
        value: `${target}`,
      },
    ],
  },
})
