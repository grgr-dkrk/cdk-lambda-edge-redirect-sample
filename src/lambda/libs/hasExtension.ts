import { extname } from 'path'

/**
 * check the uri has an extension.
 * @param { string } uri - uri
 * @return { boolean } boolean
 */
export const hasExtension = (uri: string): boolean => {
  const extension = extname(uri)
  return !!(extension && extension.length > 0)
}
