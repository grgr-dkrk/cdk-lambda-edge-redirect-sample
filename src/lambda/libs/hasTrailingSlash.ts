/**
 * check the uri has trailing slash.
 * @param { string } uri - uri
 * @return { boolean } boolean
 */
export const hasTrailingSlash = (uri: string): boolean => uri.endsWith('/')
