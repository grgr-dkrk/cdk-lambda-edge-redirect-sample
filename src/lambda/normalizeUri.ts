import { hasExtension, hasTrailingSlash } from './libs'

/**
 * normalizeUri
 * @param { string } uri - URI
 * @return { string } newUri - new URI
 */
export const normalizeUri = (uri: string): string => {
  const INDEX_PATH = 'index.html' as const

  // if uri has some extension, return uri as it is.
  if (hasExtension(uri)) return uri

  // return `${uri}/index.html`
  return hasTrailingSlash(uri) ? `${uri}${INDEX_PATH}` : `${uri}/${INDEX_PATH}`
}
