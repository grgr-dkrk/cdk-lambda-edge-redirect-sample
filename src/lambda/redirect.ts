import { CloudFrontResultResponse } from 'aws-lambda'
import { createRedirectResponse } from './libs'

/**
 * redirect
 * @param { string } uri - URI
 * @return { Response | null } response | null - Response for Redirect or Null
 */
export const checkRedirectResponse = (
  uri: string,
): CloudFrontResultResponse | null => {
  // if uri is /foo/*, redirect to /bar/.
  if (uri.match(/^\/foo\/?/)) {
    return createRedirectResponse('/bar/')
  }

  return null
}
