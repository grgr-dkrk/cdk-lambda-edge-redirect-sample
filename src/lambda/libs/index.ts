import { hasExtension } from './hasExtension'
import { hasTrailingSlash } from './hasTrailingSlash'
import { createRedirectResponse } from './createResponse'

export { createRedirectResponse, hasExtension, hasTrailingSlash }
