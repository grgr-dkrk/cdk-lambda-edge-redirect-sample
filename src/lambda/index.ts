import { CloudFrontRequestEvent, CloudFrontRequestResult } from 'aws-lambda'
import { normalizeUri } from './normalizeUri'
import { checkRedirectResponse } from './redirect'

export const lambdaFunctions = async (
  event: CloudFrontRequestEvent,
): Promise<CloudFrontRequestResult> => {
  const request = event.Records[0].cf.request
  const { uri } = request

  // redirect
  const redirectResponse = checkRedirectResponse(uri)
  if (redirectResponse) {
    return redirectResponse
  }

  // rewrite
  const normalizedUri = normalizeUri(uri)
  return { ...request, uri: normalizedUri }
}
