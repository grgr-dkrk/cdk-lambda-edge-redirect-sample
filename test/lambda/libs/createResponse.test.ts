import { createRedirectResponse } from '../../../src/lambda/libs'

describe('createRedirectResponse', () => {
  it('snapshot: createRedirectResponse', () => {
    expect(createRedirectResponse('https://dkrk-blog.net')).toMatchSnapshot()
  })
  describe('status code', () => {
    it('status code: 301', () => {
      const { status } = createRedirectResponse('https://dkrk-blog.net', '301')
      expect(status).toBe('301')
    })
    it('status code: 302', () => {
      const { status } = createRedirectResponse('https://dkrk-blog.net', '302')
      expect(status).toBe('302')
    })
    it('status code: 307', () => {
      const { status } = createRedirectResponse('https://dkrk-blog.net', '307')
      expect(status).toBe('307')
    })
  })
  describe('status description', () => {
    it('status description: something', () => {
      const { statusDescription } = createRedirectResponse(
        'https://dkrk-blog.net',
        '301',
        'moved',
      )
      expect(statusDescription).toBe('moved')
    })
    it('status description: none', () => {
      const { statusDescription } = createRedirectResponse(
        'https://dkrk-blog.net',
      )
      expect(statusDescription).toBe(undefined)
    })
  })
})
