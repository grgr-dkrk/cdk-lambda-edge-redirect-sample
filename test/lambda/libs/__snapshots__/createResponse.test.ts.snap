// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRedirectResponse snapshot: createRedirectResponse 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "https://dkrk-blog.net",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;
